### PR TITLE
fix(postgres,mysql): better support for enum and enum array

### PR DESCRIPTION
**Pull request by vlapo**
- better support all types of typescript enums (numeric, string, heterogeneous) + tests + docs - Fix https://github.com/typeorm/typeorm/issues/3393, Fix https://github.com/typeorm/typeorm/issues/3072
```typescript

export enum NumericEnum {
    ADMIN,
    EDITOR,
    MODERATOR,
    GHOST
}

export enum StringEnum {
    ADMIN = "a",
    EDITOR = "e",
    MODERATOR = "m",
    GHOST = "g"
}

export enum HeterogeneousEnum {
    NO = 0,
    YES = "YES",
}
```
 - default values for enum array coluns (postgres) could be now defined as array - Fix https://github.com/typeorm/typeorm/issues/3323
```typescript
@Column({
    type: "enum",
    enum: StringEnum,
    array: true,
    default: [StringEnum.ADMIN]
})
stringEnums: StringEnum[];
```
 - add enum type to sample - Fix https://github.com/typeorm/typeorm/issues/3392

### CODE REVIEW

File: src/driver/mysql/MysqlDriver.ts [hunk 1]
```diff
@@ -471,6 +473,9 @@ export class MysqlDriver implements Driver {
 
         } else if (columnMetadata.type === "simple-json") {
             value = DateUtils.stringToSimpleJson(value);
+
+        } else if (columnMetadata.type === "enum" && !isNaN(value)) {
```
Comment (pleerock) [inline src/driver/mysql/MysqlDriver.ts]: what if value was defined as `"1"` and user expects numbers?

Comment (pleerock) [reply src/driver/mysql/MysqlDriver.ts]: I don't think we'll be able to make "heterogeneous" way possible to work. The only option I see is to support string-only or number-only. If enum was defined in a   "heterogeneous" way - we either throw error either enforce strings.

(to understand if its string or numeric based we do `Object.keys(enum).every(key => typeof enum[key] === "number")`)

**Comment by pleerock** [comment]
This PR is amazing~

File: src/driver/mysql/MysqlDriver.ts (hunk 1 above)
Comment (vlapo) [reply src/driver/mysql/MysqlDriver.ts]: Yeah I decided to force all string numbers to classic javascript numbers, so values of this enum will be transformed to number after select from databases:
```typescript
enum TestEnum {
    NO = "0",
    YES = "1"
}
```
But in line you are highlighted we can check if exists numeric enum value (`Object.keys(enum).some(key => enum[key] === parseInt(value))`) and convert it only if exists. After that change the only one edge case will be enum like this:
```typescript
enum TestEnum {
    YES1 = 1
    YES = "1"
}
```
And this looks like really edge case.

Comment (pleerock) [reply src/driver/mysql/MysqlDriver.ts]: Right, last case is really edge case and I think its okay if we don't cover it. But first case should work properly (convert "0" to string, not number).

Comment (vlapo) [reply src/driver/mysql/MysqlDriver.ts]: Updated src and added test enum with numeric values as string.

**Comment by Noob7** [comment]
Doesnt work on mysql and its 2023. What am i doing wrong? 
  @Exclude()
  @Column({
    type: 'enum',
    enum: UserRole,
    array: true,
  })
  roles: UserRole;
